### ISSUE TITLE
Missing freeze on fold: (urem X, -1) -> select(X == -1, 0, x)

### ISSUE THREAD

**Issue by RKSimon** [llvm/llvm-project#55291]
https://alive2.llvm.org/ce/z/j9ZLMJ
```
----------------------------------------
define i8 @src(i8 %a0) {
  %r = urem i8 %a0, 255
  ret i8 %r
}
define i8 @tgt(i8 %a0) {
  %c = icmp eq i8 %a0, 255
  %r = select i1 %c, i8 0, i8 %a0
  ret i8 %r
}
Transformation doesn't verify!

----------------------------------------
define i8 @src(i8 %a0) {
  %r = urem i8 %a0, 255
  ret i8 %r
}
define i8 @tgt(i8 %a0) {
  %f = freeze i8 %a0
  %c = icmp eq i8 %f, 255
  %r = select i1 %c, i8 0, i8 %f
  ret i8 %r
}
Transformation seems to be correct!
```
Noticed while working on Issue #55271

**Comment by RKSimon** [comment]
The DAG version of the fold has the same issue

**Comment by rotateright** [comment]
Fixed with d428f09b2c9d
